HFE.md

```diff
@@ -0,0 +1,632 @@
+# Hash Items TTL Support in Valkey RFC
+
+
+
+## Overview
+
+The open-source Valkey community, recognized the growing need for fine-grained control over hash object metadata by introducing Hash Field Expiration (HFE), a feature that allows setting time-to-live (TTL) for individual fields within hash objects. Driven by customer requests for more precise data management and inspired by Redis Ltd.'s recent introduction, We introduce an approach focusing on memory efficiency and performance while maintaining existing user experiences. The solution involves implementing two expiration mechanisms: lazy expiration, which releases fields when accessed, and a background cron job that scans and expires volatile objects. By developing an efficient active expiry mechanism, we aim to reduce cron cycle times and minimize tail latency, ultimately providing developers with granular control over data lifecycle management. The implementation also prioritizes API compatibility with existing Redis clients, enabling seamless migration and adoption for users seeking more sophisticated hash object management capabilities. 
+
+
+## Requirements 
+
+* Support Redis compatible API to set, get and manipulate Hash field TTL.
+* Support both Lazy and active expiry mechanisms for Hash field expirations.
+* Support Replication of elements TTL as well as expired element replication.
+* Extended support for the same functionality with Sets and Sorted sets.
+
+### Tenets
+
+* **memory efficiency** - At the highest priority we target a minimal metadata overhead used in order to store and manage items TTL. While the optimal overhead to maintain item TTL is 8 bytes (could be less if we allow keeping offsets from the existing epoch diff time), we understand that maintaining active expiry logic will require use of more bytes for metadata. We will make our top priority effort to minimize this overhead.
+* **Latency** - After memory efficiency considerations we will require a solution which provides low latency for hash operations. Hash objects are expected to operate in O(1) for all single access operations (get, set, delete etc...) and we will not break this promise even for items with expiry.
+* **CPU efficiency** - After latency we priorities system CPU efficiency. For example we would like to avoid high CPU utilization caused by need to perform null active expiry checks during cron runs.
+* **Copherency** -  We would like the reported system state to match the logical state as much as possible. For example the reported number of keys per DB is expected to match the number of keys which are NOT logically expired.  
+
+## Interface Requirements
+
+### New command API
+
+The proposed API is very much identical to the Redis provided API (Redis 7.4 + 8.0). This is intentionally proposed in order to avoid breaking client applications already opted to use hash items TTL.   
```

Comment:
I don't know if KeyDB's API is the same. If it's different, then we should consider supporting it and explain why we don't if we don't.

Reply:
I can look into it. I made an attempt to follow the Redis API since I guess most clients would align to it.